x/mongo/driver/topology/rtt_monitor.go

```diff
@@ -179,20 +183,41 @@ func (r *rttMonitor) runHellos(conn *connection) {
 	}
 }
 
-// reset sets the average and min RTT to 0. This should only be called from the server monitor when an error
-// occurs during a server check. Errors in the RTT monitor should not reset the RTTs.
+// reset sets the average, min, and stddev RTT to 0. This should only be called from the server monitor
+// when an error occurs during a server check. Errors in the RTT monitor should not reset the RTTs.
 func (r *rttMonitor) reset() {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 
 	r.movingMin = list.New()
 	r.averageRTT = 0
 	r.averageRTTSet = false
+	r.stddevSum = 0
+	r.callsToAppendMovingMin = 0
+}
+
+func (r *rttMonitor) calcStddev() float64 {
```

Comment:
Suggest generalizing this as a function and putting into an internal/stats package in case we ever need to use it again. We should call it something like `StandardDeviationList`. We should (albeit unnecessary for this case) to ensure [numerical stability](https://mathworld.wolfram.com/NumericalStability). That is, I suggest implementing an [algorithm](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm) that does not require calculating the mean and variance in separate loops:

```go
func StandardDeviationList[T time.Duration | float64](l *list.List) float64 {
	if l.Len() == 0 {
		return 0
	}

	var mean, variance float64
	count := 0.0

	for el := l.Front(); el != nil; el = el.Next() {
		count++
		sample := float64(el.Value.(T))

		delta := sample - mean
		mean += delta / count
		variance += delta * (sample - mean)
	}

	return math.Sqrt(variance / count)
}
```

Here is a template for testing this:

```go
func Test_calcStddev_Float64(t *testing.T) {
	tests := []struct {
		name string
		data []float64
		want float64
	}{
		{
			name: "multiple",
			data: []float64{-1, 1, 6, 8, 10},
			want: 4.1665333,
		},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			l := list.New()
			for _, d := range test.data {
				l.PushBack(d)
			}

			got := StandardDeviationList[float64](l)

			assert.InDelta(t, test.want, got, 1e-6)
		})
	}
}

func Test_calcStddev_Duration(t *testing.T) {
	tests := []struct {
		name string
		data []time.Duration
		want float64
	}{
		{
			name: "multiple",
			data: []time.Duration{
				time.Millisecond,
				2 * time.Millisecond,
				time.Microsecond,
			},
			want: 816088.36667497,
		},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			l := list.New()
			for _, d := range test.data {
				l.PushBack(d)
			}

			got := StandardDeviationList[time.Duration](l)

			assert.InDelta(t, test.want, got, 1e-6)
		})
	}
}
```

Reply:
So cool finding an algorithm that can calculate variance in one pass!! Thank you!